compiler/src/dmd/dmodule.d

```diff
@@ -358,6 +358,7 @@ extern (C++) final class Module : Package
     FileType filetype;          // source file type
     bool hasAlwaysInlines;      // contains references to functions that must be inlined
     bool isPackageFile;         // if it is a package.d
+    Edition edition;            // language edition that this module is compiled with
```

Comment:
IMO this should be part of `Package` not `Module` and should be settable by an annotation in the `package.d` file that affects all sub-packages (unless overwritten) and all sibling files.

The less configuration the user must specify the better. Note that since `Module` inherits from `Package` we can still cache the result of the calculation of `edition` in a call to `hasEdition`.

I understand this is a first step, so I fine with this as is, for now, but we should definitely implement that at some point

Reply:
Not every package will have a package.d. I'm also not convinced that every module in a package must be the same edition. We should be able to have modules with different editions able to work together (if possible).

Reply:
> I'm also not convinced that every module in a package must be the same edition. 

That will still work: an edition specified on a module is checked first, then a sibling `package.d` (if it exists) then `package.d` files up the module hierarchy, then whatever the default package is (oldest/specified on cmd line/whatever).

Reply:
That would mean importing module `a.b.c.d.e.f.g.h` would have to open all parent modules too to check for package files and their edition version.

Reply:
first I doubt anyone nests that deep anyway, second if they exist they are probably already on the command line, and third that lookup will be cached